File: framework/scala/src/main/scala/soda/scala/web/setup/SetupHandler.scala
package soda.scala.web.setup

import com.sun.net.httpserver.HttpExchange
import soda.scala.web.{BaseHandler, BusinessCode, Logger, ParameterMissingException, ServiceException}
import soda.scala.web.http.RequestHelper
import soda.scala.web.resp.{Response, ResponseFactory}

class SetupHandler extends BaseHandler {
  override def doPost(exchange: HttpExchange): Response = {
    // multipart/form-data, key = ?, jar = ?
    val data = RequestHelper.multipartFormData(exchange)

    val key = data.firstValue("key") match {
      case Some(key) => key
      case None => throw new ParameterMissingException("key")
    }

    val bytes = data.firstFile("jar") match {
      case Some(bytes) => bytes
      case None => throw new ParameterMissingException("jar")
    }

    ClassLoaderCache.setupForJar(key, bytes)
    Logger.info(s"reset class loader for key $key")
    ResponseFactory.success()
  }

}

File: framework/kotlin/src/main/kotlin/soda/kotlin/web/setup/SetupHandler.kt
package soda.kotlin.web.setup

import com.sun.net.httpserver.HttpExchange
import soda.kotlin.web.BaseHandler
import soda.kotlin.web.Logger
import soda.kotlin.web.exception.ParameterMissingException
import soda.kotlin.web.http.RequestHelper
import soda.kotlin.web.resp.Response
import soda.kotlin.web.resp.ResponseFactory

class SetupHandler: BaseHandler() {

    override fun doPost(exchange: HttpExchange?): Response {
        // multipart/form-data, key = ?, jar = ?
        val data = RequestHelper.multipartFormData(exchange!!)

        val key = data.firstValue("key") ?: throw ParameterMissingException("key")
        val bytes = data.firstFile("jar") ?: throw ParameterMissingException("jar")

        ClassLoaderCache.setupForJar(key, bytes)
        Logger.info("reset class loader for key $key")
        return ResponseFactory.success()
    }

}